Flaky test: CardBrowserViewModelTest - `suspend - notes - some notes suspended `

macOS runner: https://github.com/ankidroid/Anki-Android/actions/runs/9357727049/job/25758121561?pr=16244

```
CardBrowserViewModelTest > suspend - notes - some notes suspended FAILED
    java.lang.AssertionError: ChangeManager: expected 1 calls
    Expected: <1>
         but: was <0>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at com.ichi2.testutils.TestChangeSubscriberKt.ensureOpsExecuted(TestChangeSubscriber.kt:50)
        at com.ichi2.anki.browser.CardBrowserViewModelTest$suspend - notes - some notes suspended$1.invokeSuspend(CardBrowserViewModelTest.kt:441)
        at com.ichi2.anki.browser.CardBrowserViewModelTest$suspend - notes - some notes suspended$1.invoke(CardBrowserViewModelTest.kt)
        at com.ichi2.anki.browser.CardBrowserViewModelTest$suspend - notes - some notes suspended$1.invoke(CardBrowserViewModelTest.kt)
        at com.ichi2.anki.browser.CardBrowserViewModelTest$runViewModelNotesTest$1.invokeSuspend(CardBrowserViewModelTest.kt:489)
        at com.ichi2.anki.browser.CardBrowserViewModelTest$runViewModelNotesTest$1.invoke(CardBrowserViewModelTest.kt)
        at com.ichi2.anki.browser.CardBrowserViewModelTest$runViewModelNotesTest$1.invoke(CardBrowserViewModelTest.kt)
        at com.ichi2.testutils.TestClass$runTest$1$1.invokeSuspend(TestClass.kt:240)
        at com.ichi2.testutils.TestClass$runTest$1$1.invoke(TestClass.kt)
        at com.ichi2.testutils.TestClass$runTest$1$1.invoke(TestClass.kt)
        at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$2$1$1.invokeSuspend(TestBuilders.kt:316)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
        at kotlinx.coroutines.test.TestDispatcher.processEvent$kotlinx_coroutines_test(TestDispatcher.kt:24)
        at kotlinx.coroutines.test.TestCoroutineScheduler.tryRunNextTaskUnless$kotlinx_coroutines_test(TestCoroutineScheduler.kt:99)
        at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt$runTest$2$1$workRunner$1.invokeSuspend(TestBuilders.kt:322)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:277)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:95)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:69)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:48)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
        at kotlinx.coroutines.test.TestBuildersJvmKt.createTestResult(TestBuildersJvm.kt:10)
        at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest-8Mi8wO0(TestBuilders.kt:310)
        at kotlinx.coroutines.test.TestBuildersKt.runTest-8Mi8wO0(Unknown Source)
        at kotlinx.coroutines.test.TestBuildersKt__TestBuildersKt.runTest-8Mi8wO0(TestBuilders.kt:168)
        at kotlinx.coroutines.test.TestBuildersKt.runTest-8Mi8wO0(Unknown Source)
        at com.ichi2.testutils.TestClass.runTest(TestClass.kt:238)
        at com.ichi2.testutils.TestClass.runTest$default(TestClass.kt:227)
        at com.ichi2.anki.browser.CardBrowserViewModelTest.runViewModelNotesTest(CardBrowserViewModelTest.kt:472)
        at com.ichi2.anki.browser.CardBrowserViewModelTest.runViewModelNotesTest$default(CardBrowserViewModelTest.kt:467)
        at com.ichi2.anki.browser.CardBrowserViewModelTest.suspend - notes - some notes suspended(CardBrowserViewModelTest.kt:438)
```

--- a/AnkiDroid/src/test/java/com/ichi2/testutils/TestChangeSubscriber.kt
+++ b/AnkiDroid/src/test/java/com/ichi2/testutils/TestChangeSubscriber.kt
@@ -16,6 +16,7 @@
 
 package com.ichi2.testutils
 
+import anki.collection.OpChanges
 import com.ichi2.libanki.ChangeManager
 import com.ichi2.libanki.undoableOp
 import timber.log.Timber
@@ -25,14 +26,13 @@ import kotlin.test.fail
  * Ensures no calls to [ChangeManager.notifySubscribers] via [undoableOp]
  */
 suspend fun ensureNoOpsExecuted(block: suspend () -> Unit) {
-    var changes = false
-    ChangeManager.subscribe { _, _ ->
-        Timber.d("ChangeManager op detected")
-        changes = true
-    }
+    val subscription = ChangeCounter()
+    ChangeManager.subscribe(subscription)
     block()
-    // we should be fine to not cleanup here
-    if (!changes) return
+    if (!subscription.hasChanges) {
+        Timber.d("ensureNoOpsExecuted: success")
+        return
+    }
 
     fail("ChangeManager should not be called; ${ChangeManager.subscriberCount} subscribers")
 }
@@ -41,15 +41,26 @@ suspend fun ensureNoOpsExecuted(block: suspend () -> Unit) {
  * Ensures no calls to [ChangeManager.notifySubscribers] via [undoableOp]
  */
 suspend fun ensureOpsExecuted(count: Int, block: suspend () -> Unit) {
-    var changes = 0
-    ChangeManager.subscribe { _, _ ->
-        Timber.d("ChangeManager op detected")
-        changes++
-    }
-    Timber.v("Listening for ChangeManager ops")
-    block()
-    if (changes == count) return
+    val subscription = ChangeCounter()
 
-    // we should be fine to not cleanup here, as the subscriber goes out of scope
+    Timber.v("Listening for ChangeManager ops")
+    ChangeManager.subscribe(subscription)
+    block()
+    if (subscription.changeCount == count) {
+        Timber.d("ensureOpsExecuted: success")
+        return
+    }
+
     fail("ChangeManager: expected $count calls; ${ChangeManager.subscriberCount} subscribers")
+}
+
+// used to ensure a strong reference to the subscription is held
+private class ChangeCounter : ChangeManager.Subscriber {
+    private var changes = 0
+    val changeCount get() = changes
+    val hasChanges get() = changes > 0
+    override fun opExecuted(changes: OpChanges, handler: Any?) {
+        Timber.d("ChangeManager op detected")
+        this.changes++
+    }
 }

